x/programs/rust/sdk-macros/src/lib.rs

```diff
@@ -40,6 +40,22 @@ pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
 
         let first_arg_err = match input.sig.inputs.first_mut() {
             Some(FnArg::Typed(PatType { ty, .. })) if is_context(ty) => {
+                let types = (context_type.as_mut(), ty.as_mut());
+
+                if let (Type::Path(context_type), Type::Path(ty)) = types {
+                    let args = [context_type, ty].map(|type_path| {
+                        type_path
+                            .path
+                            .segments
+                            .last_mut()
+                            .map(|segment| &mut segment.arguments)
+                    });
+
+                    if let [Some(context_type_args), Some(ty_args)] = args {
+                        *context_type_args = ty_args.clone();
+                    }
+                }
+
                 std::mem::swap(&mut context_type, ty);
```

Comment:
I think that you can replace all of this by

```suggestion
                context_type = ty.clone();
```

That is going to preserve the explicit generic passed to the `Context` struct if it is needed to not be `()`